lib/iris/fileformats/netcdf.py

```diff
@@ -1825,166 +1825,57 @@ def _get_coord_variable_name(self, cube, coord):
-        * dimension_names (list):
-            Names for each dimension of the cube.
-        * ancillary_variable (:class:`iris.coords.AncillaryVariable`):
-            The ancillary variable to be saved to the CF-netCDF file.
-
-        Returns:
-            The string name of the associated CF-netCDF variable saved.
+        given dimensional_metadata.
 
-        """
-        # Note: currently shares variable creation code with cell-measures.
-        return self._inner_create_cf_cellmeasure_or_ancil_variable(
-            cube, dimension_names, ancillary_variable
-        )
-
-    def _create_cf_coord_variable(self, cube, dimension_names, coord):
-        """
-        Create the associated CF-netCDF variable in the netCDF dataset for the
-        given coordinate. If required, also create the CF-netCDF bounds
-        variable and associated dimension.
+        ..note::
+            If the metadata element is a coord, it may also contain bounds.
+            In which case, an additional var is created and linked to it.
 
         Args:
 
         * cube (:class:`iris.cube.Cube`):
             The associated cube being saved to CF-netCDF file.
-        * dimension_names (list):
-            Names for each dimension of the cube.
-        * coord (:class:`iris.coords.Coord`):
-            The coordinate to be saved to CF-netCDF file.
+        * cube_dim_names (list of string):
+            The name of each dimension of the cube.
+        * element:
+            An Iris :class:`iris.coords.DimensionalMetadata`, belonging to the
+            cube.  Provides data, units and standard/long/var names.
 
         Returns:
-            The string name of the associated CF-netCDF variable saved.
+            var_name (string):
+                The name of the CF-netCDF variable created.
 
         """
-        cf_name = self._get_coord_variable_name(cube, coord)
+        # Work out the var-name to use.
+        cf_name = self._get_coord_variable_name(cube, element)
         while cf_name in self._dataset.variables:
             cf_name = self._increment_name(cf_name)
 
-        # Derive the data dimension names for the coordinate.
-        cf_dimensions = [
-            dimension_names[dim] for dim in cube.coord_dims(coord)
-        ]
-
-        if np.issubdtype(coord.points.dtype, np.str_):
-            string_dimension_depth = coord.points.dtype.itemsize
-            if coord.points.dtype.kind == "U":
+        # Get the list of file-dimensions (names), to create the variable.
+        element_dims = [
+            cube_dim_names[dim] for dim in element.cube_dims(cube)
+        ]  # NB using 'cube_dims' as this works for any type of element
+
+        # Get the data values, in a way which works for any element type, as
+        # all are subclasses of _DimensionalMetadata.
+        # (e.g. =points if a coord, =data if an ancillary, etc)
+        data = element._values
+
+        if np.issubdtype(data.dtype, np.str_):
```

Comment:
@pp-mo This is potentially new behaviour for `cell-measures` and `ancillary-variables`, right?

Reply:
Potentially yes.
This code is now unified usage for coordinates, cell-measures and ancillaries (and in future will be applied to connectivities and mesh-location coordinates).
From unifying the code, there is in fact a fair bit of code here that used to be applied only to coordinates but now can trigger with other types of things, so we need to consider that carefully on a case-by-case basis.

In this particular case, I think string-typed  ancillaries are a reasonable possibility, and this ought to handle them OK.
I don't think string cell-measures are at all likely.
However I _think_ the logic here is pretty general : when you have a array of data to process, and it might sometimes by strings.

Other cases like this :
  * units calendar handling : see comment 
  * the odd sections I made special to dim-coords : should be pretty safe.
  * special handling of STASH attributes : could apply to AuxCoords in some cases, I think.  
    We could narrow the scope when this is applied, but I don't think it's very dangerous as-is.

Reply:
@pp-mo Cool. I guess we'll see if use cases arrive that exercise this, but at this moment in time I'm not worried if you're not.

I just wanted to confirm that this would be new behaviour for `cell-measures` and `ancillary-variables`, thanks for clarifying :+1: